### PR TITLE
Improving Lancaster stemmer with strip prefix function and customizable rules from Whoosh

### PR DESCRIPTION
**Pull request by jaehoonhwang**
Issue: #1648 

- Clean up documentation
- Added doctests (copied from `lancaster.py`)
- Added AUTHORS.md
- Fix the code where if there is no rule table present in __init__, the stemmer uses default rule.

Code is from [Whoosh](https://bitbucket.org/mchaput/whoosh/src/e344fb64067e45d47ec62dc65a75a50be51264a7/src/whoosh/lang/paicehusk.py?at=default&fileviewer=file-view-default)

If I could get more detail on `nltk` testing, that would be great.
Because, I'm a bit confused with doctest and how to run the test.

### CODE REVIEW

**Comment by alvations** [comment]
@jaehoonhwang the tests are failing because the stemming function is returning empty strings instead of the expected stems. https://nltk.ci.cloudbees.com/job/pull_request_tests/118/TOXENV=py35-jenkins,jdk=jdk8latestOnlineInstall/testReport/junit/nltk.stem/paicehusk/PaiceHuskStemmer/

Could you take a look at what's wrong? Or would you need some guidance on debugging it?

**Comment by jaehoonhwang** [comment]
Oh no. I already verified it.
When there is no stem, such as word `multiply` the function returns an empty string.
I was wondering:
1. How can I run doctest locally? Should I use following line?

```py
if __name__ == "__main__":
    import doctest
    doctest.mod()
```

2. Is doctest enough for this module or do I need to write `unittest`?

**Comment by alvations** [comment]
When no pattern matches, possibly the normal way to handling that is to return the original string instead of an empty string. E.g.

```
>>> from nltk.stem import PorterStemmer
>>> porter = PorterStemmer()
>>> porter.stem('dog')
'dog'
```

Could you make changes to the code such that it returns the original string when there are no stem pattern match?

File: nltk/stem/paicehusk.py
Comment (alvations) [inline nltk/stem/paicehusk.py]: The comments right for the lancaster stemmer but is Paice-Husk stemmer exhibiting the same behavior?

**Comment by jaehoonhwang** [comment]
Yes.
I was working on it, just in case you wanted that way.

File: nltk/stem/paicehusk.py
Comment (jaehoonhwang) [reply nltk/stem/paicehusk.py]: Yes. When I tested them locally, the doctest didnt give me any error except for empty string ones. But I copied lancaster doctest for testing purpose. I will checl them in the morning

File: nltk/stem/paicehusk.py
Comment (alvations) [inline nltk/stem/paicehusk.py]: It'll be good to add the ruletable variable to the stemmer object that way, it can be modified later on (if needed), e.g. 

```python
self._ruletable = ruletable if ruletable else self.defaultrules
self.read_rules(self._ruletable)
```

File: nltk/stem/paicehusk.py
Comment (alvations) [inline nltk/stem/paicehusk.py]: I think this `append` here don't change the `self.rules` and it isn't used anywhere else in this function, right? Would it break anything if we remove this?

File: nltk/stem/paicehusk.py
Comment (alvations) [inline nltk/stem/paicehusk.py]: I think if we are not modifying `self.rule_expr` and `self.rules`, we might not need to have a local copy of this =)

File: nltk/stem/paicehusk.py
Comment (alvations) [inline nltk/stem/paicehusk.py]: Interesting, any idea why `strip_prefix` is necessary? @ExplodingCabbage any tips?

File: nltk/stem/paicehusk.py
Comment (alvations) [inline nltk/stem/paicehusk.py]: This if-condition `if intac and not is_intact` seems a little funny. Maybe some comments on this would help users and us better maintain the code

File: nltk/stem/paicehusk.py
Comment (alvations) [reply nltk/stem/paicehusk.py]: Cool! When we have more of them maybe we can add a wiki on different effects of different stemmers =)

**Comment by alvations** [comment]
Thanks @jaehoonhwang! I've finished the review, I hope the line-by-line review helps the revision =)

**Comment by ExplodingCabbage** [comment]
Hmm. Licensing aside (though it would be good if someone could confirm there are no issues there, too), how do Whoosh feel about NLTK copying their code? To avoid code duplication and allow any future fixes and features to benefit both userbases, would it good for them to make NLTK a dependency and import the stemmer from it, or perhaps for NLTK to extract out this particular stemmer into a standalone package that both Whoosh and NLTK can depend upon? I believe @mchaput is the maintainer of Whoosh; I suggest that it would be both useful and courteous to give him a chance to weigh in here and at https://github.com/nltk/nltk/issues/1648 before merging copies of any of his stuff.

Also, if we're going to copy people's work, I think they should be credited in AUTHORS.md. You may want to emulate the approach I took with the Porter Stemmer.

I'll take a further look at this PR when I have the chance, probably at the weekend.

**Comment by alvations** [comment]
I agree with @ExplodingCabbage on asking the Whoosh maintainers before we merge the code.  Started a thread on https://groups.google.com/forum/#!topic/whoosh/8o_iYCn1Sao

**Comment by jaehoonhwang** [comment]
@alvations Is there any more changes you want me to do?

**Comment by ExplodingCabbage** [comment]
- [x] The stemmer should be imported in `nltk/stem/__init__.py` like all the other stemmers are.

File: nltk/stem/paicehusk.py
Comment (alvations) [inline nltk/stem/paicehusk.py]: This would be a more explicit example:

```python
paicehusk = PaiceHuskStemmer()
stemmed_word = paicehusk.stem(word)
```

File: nltk/stem/paicehusk.py
Comment (alvations) [inline nltk/stem/paicehusk.py]: It'll be good to give an example of how to instantiate `my_rules_string`, esp. the syntax of how `my_rules_string` should be written.

Comment (jaehoonhwang) [reply nltk/stem/paicehusk.py]: Ok, I'm on it.

File: nltk/stem/paicehusk.py
Comment (alvations) [inline nltk/stem/paicehusk.py]: Missing words?

**Comment by alvations** [comment]
For now there's no preference for doctest or unittest but writing doctest for the stemmers will be a good start.

BTW, after reading the literature and the code, the Paice-Husk stemmer has the same reference as the Lancaster stemmer and the rules and algorithm looks really similar. **Does anyone know whether they're the same thing?** 

If they are the same, then merging some code to allow custom rules from here to the `nltk.stem.lancaster.py` would be more appropriate than keeping two copies of the same algorithm.

Sorry I didn't realize that they might be the same earlier.

**Comment by ExplodingCabbage** [comment]
> BTW, after reading the literature and the code, the Paice-Husk stemmer has the same reference as the Lancaster stemmer and the rules and algorithm looks really similar. Does anyone know whether they're the same thing?

I've also been trying to figure that out, and I think the answer is "yes". In particular, NLTK's `lancaster.py` cites

> Paice, Chris D. "Another Stemmer." ACM SIGIR Forum 24.3 (1990): 56-61.

which is *also* the paper that many sources on the internet cite as the original definition of Paice/Husk. Note also that Chris Paice worked at Lancaster University before he died (per http://wp.lancs.ac.uk/chrispaice/).

I have no idea why the word "Husk" appears in the name - that bit of history yet eludes me.

Also, unfortunately the paper only seems to exist on the internet behind a paywall, SciHub seems unable to pirate it for me, and I can't ask Chris directly for it because he's dead. Perhaps there are some academics here who have access via their institutions?

I can, however, test purported implementations of Paice/Husk on the internet against this one. For instance, there's one at http://paicehusk.appspot.com/. It doesn't have the prefix-removal stuff that the one in this PR does (e.g. this PR stems `'kilogram'` to `'gram'`; that implementation stems it to `'kilogram'`). `lancaster.py` also doesn't do that prefix removal. That strongly hints that `lancaster.py` is the more faithful implementation.

**Comment by ExplodingCabbage** [comment]
Aha, "Husk" seems to be the surname of another researcher that Chris Paice worked worth - this article is apparently by *C.D. Paice* and *G.D. Husk*: http://www.sciencedirect.com/science/article/pii/0885230887900039

I'm guessing that when someone gets their hands on a PDF of "Another Stemmer" we'll learn that Husk was a co-author, and that through some quirk of history his name never gets included when academics cite the paper and yet made it into the popular name of the stemmer?

**Comment by alvations** [comment]
Ah, Paice wrote and published the paper and Husk wrote the software, and they're both from lancaster: https://pdfs.semanticscholar.org/3b8b/099f62391ff24e94b10ebafbb0488be83632.pdf and https://web.archive.org/web/20060501021257/http://www.comp.lancs.ac.uk/computing/research/stemming/Links/implementations.htm

It looks like the prefix stripping function is whoosh's addition, the [original pascal code](https://web.archive.org/web/20060501021257/http://www.comp.lancs.ac.uk/computing/research/stemming/Files/Pascal.zip) didn't have that. 

@jaehoonhwang I think it'll be better if we merge the custom rules code from here to lancaster and allow an option to allow users to  remove prefix or keep the original implementation.

**Comment by ExplodingCabbage** [comment]
Yeah, he's from Lancaster, per his Facebook: https://www.facebook.com/ghusk

**Comment by jaehoonhwang** [comment]
I can remove prefix function, if you prefer.

**Comment by alvations** [comment]
@jaehoonhwang we should take what whoosh has (in `nltk/stem/paicehusk.py`) and then add then to the `nltk/stem/lancaster.py` code. That's so that we don't maintain 2 variants of the same algorithm at separate places. E.g. for Porter stemmer, if i'm not mistaken, @ExplodingCabbage allowed users to choose the variants, see https://github.com/nltk/nltk/blob/develop/nltk/stem/porter.py#L49

There're some good things that whoosh has that the current lancaster stemmer in NLTK does, more specifically, the:

 - prefix stripping function
 - custom rule

**Comment by jaehoonhwang** [comment]
Those functionalities are now added to Lancaster and I removed `paicehusk.py`.
~~Can I get a rule for testing custom rule tuples?~~
~~I've tried different things, but can't come up with anything that works.~~
I got it.

File: nltk/stem/lancaster.py
Comment (alvations) [inline nltk/stem/lancaster.py]: It'll be good to acknowledge the original src of this function from whoosh here by adding a docstring that includes the link: https://bitbucket.org/mchaput/whoosh/src/e344fb64067e45d47ec62dc65a75a50be51264a7/src/whoosh/lang/paicehusk.py#paicehusk.py-73

Comment (ExplodingCabbage) [reply nltk/stem/lancaster.py]: Better yet, let's also try to track down the history of how it got added to Whoosh via their source control. For all we know it may have been a copy-and-paste to begin with from some other source that we're failing to credit! Knowing the history will also let us credit individual authors in CONTRIBUTORS.md.

None of this is vital, but would be nice (and also potentially *practically* useful in helping understand the alterations that were made to the original algorithm and what motivated them).

Comment (alvations) [reply nltk/stem/lancaster.py]: This is interesting, it seems like many versions of Porter stemmer (except NLTK's) contain that function ;P

https://www.google.com.sg/search?client=safari&rls=en&q=for+prefix+in+(%22kilo%22,+%22micro%22,+%22milli%22,+%22intra%22,+%22ultra%22,+%22mega%22

File: nltk/stem/lancaster.py
Comment (ExplodingCabbage) [inline nltk/stem/lancaster.py]: - [x] Surely we should be coercing the word to lowercase *before* the `strip_prefix()` call, not after? Won't this implementation map `KILOGRAM` to `kilogram` but `kilogram` to `gram`?

File: nltk/stem/lancaster.py
Comment (jaehoonhwang) [reply nltk/stem/lancaster.py]: For now I'll add Whoosh as a source.

File: nltk/stem/lancaster.py
Comment (jaehoonhwang) [reply nltk/stem/lancaster.py]: Confirmed and fixed.

**Comment by jaehoonhwang** [comment]
Should I add Whoosh on `contributing.md`?
Also, do you guys think having `_strip_prefix` and `__strip_prefix` is too confusing?
One is class member and one is class function.
Also, i have an argument in Lancaster init as `strip_prefix`

- `strip_prefix`: Argument variable
- `_strip_prefix`: Class variable (bool)
- `__strip_prefix`: Class member function.

**Comment by stevenbird** [comment]
@jaehoonhwang: I agree that the variable names are too similar, given the different purposes.
I agree about acknowledging indirect contributions.

**Comment by jaehoonhwang** [comment]
Any other thing I should do? @alvations

**Comment by alvations** [comment]
@jaehoonhwang sorry for the late reply. It seems like the Whoosh maintainers didn't reply on https://groups.google.com/forum/#!topic/whoosh/8o_iYCn1Sao 

Let's wait till the weekend and if there's still no response, it looks good to merge after a last round of quick review @ExplodingCabbage and me. 

@ExplodingCabbage @mchaput anything to add?

**Comment by jaehoonhwang** [comment]
Thanks for the reply!

**Comment by alvations** [comment]
Good news, we got the go ahead to merge from @mchaput =) 

@jaehoonhwang just one small request, do a `git pull` and `git merge` to include the docstring addition that @ExplodingCabbage wrote. 

@ExplodingCabbage any thing to add to this PR?

**Comment by jaehoonhwang** [comment]
Ok. I'll do it in few hours and do all those. Thanks for getting back.

**Comment by jaehoonhwang** [comment]
@alvations I think this is it. Thank you and @ExplodingCabbage for the help!

**Comment by ExplodingCabbage** [comment]
A few things:

1. There's some commented-out code that this version currently adds that shouldn't be there
2. We've added an argument called `rule_tuples` (plural), but doesn't it actually take (at least in our test case) a *single* tuple? And shouldn't it therefore be named `rule_tuple` (singular)? Am I missing something?
3. The (pre-existing) comment here is interesting:

    ```py
        # If the user hasn't supplied any rules, setup the default rules
        if len(self.rule_dictionary) == 0:
            self.parseRules(LancasterStemmer.rule_tuple)
    ```

    It looks like the possibility of users passing in custom rules was at least something that the authors of the NLTK version of the stemmer had conceived of, despite not taking them as an argument to the constructor. I guess this is why `parseRules()` is public and takes a `rule_tuple` argument; using the *old* stemmer, I can do this (similar to the doctest added in this PR):

    ```
    >>> st_custom = LancasterStemmer()
    >>> st_custom.parseRules(rule_tuple=("ssen4>", "s1t."))
    >>> st_custom.stem("ness") # Change s to t
    'nest'
    ```

    As such, one of the two pieces of functionality this PR adds was actually already supported, just without us noticing!

    I guess I don't object to adding the `rule_tuple` argument to the constructor, though, since:
    - it doesn't break the old custom rules API
    - the old custom rules API was completely undocumented and non-obvious; the new way is significantly more discoverable

    However, we *should* delete that pre-existing comment since this PR renders it completely wrong (since the block being commented on can now run even if the user has supplied rules, and the rules it sets up are not necessarily default ones).

**Comment by jaehoonhwang** [comment]
That makes sense.
What should I do with this PR?
This seems like redundant PR that supports already existing function in Lancasterstemmer.

**Comment by jaehoonhwang** [comment]
@ExplodingCabbage 
I've fixed/added documentation on parseRules  and __stripPrefix, removed commented out codes, fixed comment in stem.
This fix still supports old API for parseRule.

**Comment by alvations** [comment]
It looks good to merge now =)

@ExplodingCabbage anything to add? BTW, just like to check, do you have merging rights?

**Review by alvations — APPROVED** [review]

**Comment by ExplodingCabbage** [comment]
> @ExplodingCabbage anything to add?

Nope. :)

> do you have merging rights?

Nope.

**Comment by jaehoonhwang** [comment]
Thank you guys for all the help!

**Comment by stevenbird** [comment]
@jaehoonhwang, @ExplodingCabbage, @alvations: thanks for wrapping this up. I'm thinking the name of this PR is out of date – perhaps one of you could fix that please, before I merge?

**Comment by stevenbird** [comment]
Thanks @alvations

**Pull request merged**
